[MLIR] Add continuous tiling to transform dialect

muneebkhan85 opened the pull request:
This patch enables continuous tiling of a target structured op using diminishing tile sizes. In cases where the tensor dimensions are not exactly divisible by the tile size, we are left with leftover tensor chunks that are irregularly tiled. This approach enables tiling of the leftover chunk with a smaller tile size and repeats this process recursively using exponentially diminishing tile sizes. This eventually generates a chain of loops that apply tiling using diminishing tile sizes.

Adds `continuous_tile_sizes` op to the transform dialect. This op, when given a tile size and a dimension, computes a series of diminishing tile sizes that can be used to tile the target along the given dimension. Additionally, this op also generates a series of chunk sizes that the corresponding tile sizes should be applied to along the given dimension.

Adds `multiway` attribute to `transform.structured.split` that enables multiway splitting of a single target op along the given dimension, as specified in a list enumerating the chunk sizes.

nicolasvasilache reviewed (commented):
Thanks for proposing to extend tiling @muneebkhan85 

Quick question, have you looked at the MultiTileSizes op (https://github.com/llvm/llvm-project/blob/1408667fdd890edf7507ae2052360de20d81c19f/mlir/test/Dialect/Linalg/multisize-tiling-full.mlir#L8).

Can the 2 approaches be (partially) merged to have less code overall?

Also adding @ftynse for visibility.

MaheshRavishankar reviewed (changes requested):
Just putting a marker cause I need to look more deeply into this. THis is a significant change and needs some time to review.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -309,6 +311,188 @@ static LogicalResult generateLoopNest(RewriterBase &rewriter, Location loc,
+/// - `CTileVector` specifies which loop nest should be continuously tiled.
+/// Note that this methods adds `scf.yield` operation for all but the innermost
+/// loop. These yield the value returned by the immediately inner tail-end loop.
+/// The caller is expected to add the scf.yield operation for all innermost
```
I have found this to be a bit of a foot gun. At some point of the implementation I did the same where the caller would apply the `scf.yield`. But that is awkward for the caller to do. Can we instead try to generate the `scf.yield` during the transformation itself?

muneebkhan85 replied:
Sorry, but this was a leftover from the previous "copied" comment. The scf.yield is actually generated for all inner loops as part of the call itself after their creation (as you suggested).

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -309,6 +311,188 @@ static LogicalResult generateLoopNest(RewriterBase &rewriter, Location loc,
+
+  // Always yield the result of the tail-end loop as this
+  // will have all the processed tiles.
+  if (!isa<func::ReturnOp>(currentLoop->getBlock()->back())) {
```
I am not sure why this is looking for a `func.return`

muneebkhan85 replied:
ok, perhaps the comments aren't that good here. It only inserts a YieldOp when there's not return at the end of the block. If there is a return then we are dealing with the outmost loop and we don't want to create a YieldOp, instead the return needs to be re-written to return the right value. The if makes sure that we are handling the first case and creating a YieldOp for the inner loops only.

MaheshRavishankar replied:
The return at the end is an artifact of the test. This seems very fragile.

muneebkhan85 replied:
I have fixed this now, so that it doesn't depend on func.return to detect the outermost loop.

MaheshRavishankar reviewed (dismissed):
Surfacing comments as I review. More to come. THis definitely is interesting, I am trying to understand the implementation. Is it possible to not duplicate all the tiling logic

nicolasvasilache reviewed (changes requested):
> Just putting a marker cause I need to look more deeply into this. THis is a significant change and needs some time to review.

Indeed it is.

I will add that we need the prerequisite cleanups (discussed [here](https://github.com/llvm/llvm-project/pull/77874#issuecomment-1911159786) and [here](https://github.com/llvm/llvm-project/pull/72178#issuecomment-1810872756)) to land first before any new code is added on these paths.

Speaking of which @MaheshRavishankar , what is the status of the required cleanup?

MaheshRavishankar commented:
> > Just putting a marker cause I need to look more deeply into this. THis is a significant change and needs some time to review.
> 
> Indeed it is.
> 
> I will add that we need the prerequisite cleanups (discussed [here](https://github.com/llvm/llvm-project/pull/77874#issuecomment-1911159786) and [here](https://github.com/llvm/llvm-project/pull/72178#issuecomment-1810872756)) to land first before any new code is added on these paths.
> 
> Speaking of which @MaheshRavishankar , what is the status of the required cleanup?

Fighting my backlog a bit. I am planning to start it in a couple of weeks.

ftynse commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -679,6 +863,276 @@ mlir::scf::tileUsingSCF(RewriterBase &rewriter, TilingInterface op,
+  std::map<int, OpFoldResult> sizesMap;
+
+  if (!options.tileSizeComputationFunction) {
+    return rewriter.notifyMatchFailure(
+        op, "missing tile size computation function");
```
These error messages are invisible unless the function is called from a pattern rewriting engine, which I suspect is not the case here. If you want to provide a better error description, you can either use `DiagnosedSilenceableFailure` from the Transform dialect or ask the caller to give an `llvm::raw_ostream`, defaulted to `llvm::nulls` that will accept the error message.

ftynse commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -679,6 +863,276 @@ mlir::scf::tileUsingSCF(RewriterBase &rewriter, TilingInterface op,
+  LLVM_DEBUG({
+    if (!forLoops.empty()) {
+      llvm::dbgs() << "LoopNest shell :\n";
+      forLoops.front()->getBlock()->dump();
```
You likely want to `->print(llvm::dbgs())` for the case when error stream and debug stream differ.

ftynse commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -679,6 +863,276 @@ mlir::scf::tileUsingSCF(RewriterBase &rewriter, TilingInterface op,
+    }
+  });
+
+  /// Remove outermost tailend loop as its only use was to compute
+  /// replacements.
+  forLoops.erase(forLoops.begin());
+
```
Not quite sure what happens here: isn't this loop produced by tiling and therefore should appear in the results?

ftynse reviewed (commented):
I left a bunch of mostly stylistic comments that I think are useful regardless of this code going in as is or not. I only commented once a particular kind of issues and there may be other issues of the same kind in the code, please fix all.

Overall, this looks quite interesting. I agree with other reviewers in my desire to see more code reuse. It is possible to alter the existing code to allow for more reuse if needed. If this is approached from the transform dialect perspective, it is possible to decompose this transformation into pieces similar to what was implemented for multisize tiling in https://github.com/llvm/llvm-project/blob/900bea9b1ce095123c03e5bb8834d8fb168378a8/mlir/test/Dialect/Linalg/multisize-tiling-full.mlir#L4. Specifically, one can separate emission on the IR computing tile sizes (before all loops since they won't change), splitting the operation into pieces of appropriate size, and tiling those pieces. I suppose that C++ code can be structured similarly and reuse the existing functions for splitting/tiling so in the end only the size computation is emitted. If a higher-level transform operation is still desired, I'd consider implementing it as a rewrite on the transform dialect itself, rather than as complex logic within C++. This was considered for multisize tiling, but not implement given the lack of interest.

muneebkhan85 commented:
> High level comment, why cant this just be an after tile and fuse peeling transformation. What is happening here seems to be peeling applied repeatedly.

An issue here is that the static tile size is created using AffineMin at the time of creating the loop nests. This makes the issue more complicated than just having to create new tail-loops and clone regions from the original loop into it.

rolfmorel commented:
Hi @ftynse, @nicolasvasilache, @MaheshRavishankar

Thank you for pointing out how multisize tiling is handled and suggesting that the continuous tiling transform can be modelled after it. 

@muneebkhan85 and I had a discussion on how continuous tiling might be possible in the transform dialect. The main issue we come up against is that rather than a statically known number of loops being generated, as with multisize tiling, continuous tiling can generate an arbitrary number of tiling loops. That is, whereas `transform.structured.multitile_sizes` returns two tile sizes and one split point, something like `transform.structured.continuous_tile_sizes` would output `n` tile sizes and `n-1` split points (based on the target size and the size of the dimension). As such we need one or more ops to encapsulate iterating over `n`, the number of splits/tiling loops.

By modelling continuous tiling after multisize tiling, we came up with the following: 

```
%linalg = transform.structured.match ops{["linalg.generic"]} in %payload
%tile_sizes, %split_points = transform.structured.continuous_tile_sizes %linalg { dimension = 0, target_size = 9} : (!transform.any_op) -> (!transform.param<i64>, !transform.param<i64>)
%head_linalg, %tail_linalgs = transform.structured.split %linalg after %split_points { dimension  = 0 }
%linalg_splits = transform.merge_handles %head_linalg, %tail_linalgs
%tiled_splits = transform.foreach %linalg_splits, %tile_sizes {
  ^bb1(%linalg_split: !transform.any_op, %tile_size: !transform.any_param):
  %tiled_linalg_split = transform.structured.tile_using_for %linalg_split [%tile_size]
  transform.yield %tiled_linalg_split
}
```

The above departs from the current transform dialect in three ways: 
 - `transform.structured.continuous_tile_sizes` is a new op with two results:
    - The first result (`%tile_sizes`) is a list containing the target size followed by all smaller powers of 2 (9, 8, 4, 2, 1 in this example)
    - The second result (`%split_points`) is a list of split points of the iteration space of the given dimension, specifying where each of the tile sizes should be applied (à la the split point of `multitile_sizes`). 
 - `transform.structured.split` is changed such that it performs a multiway split at once. The above assumes that `split` would still produce only two outputs, though the second handle (`%tail_linalgs`) would collect all but the first split-off part.
 - `transform.foreach` is changed to take multiple handles by making `target` variadic and iterates over all handles at once by "zipping" the lists associated with the handles. The enclosed block would take as many arguments as there are targets.
 
For the generalisation of `foreach` we have come across other use cases as well (in particular for fusing loops coming from co-indexed handles). 

Could you let us know if the above is similar to what you had in mind? We would be happy to work on landing the above three changes.

If there are other more preferable paths to supporting continuous tiling, do let us know. @ftynse, you mentioned an approach whereby the transform dialect would itself be subject to rewrites. Do you have a reference for us to look at?

ftynse commented:
SGTM in principle. I'm not quite sure how this tiling would work for multiple dimensions though.

> transform.structured.split is changed such that it performs a multiway split at once. The above assumes that split would still produce only two outputs, though the second handle (%tail_linalgs) would collect all but the first split-off part.

Nit: I'd rather just have one result, and your example merges the two results anyway. If the two distinct results are necessary, we can either use `transform.split_handle` or have an attribute controlling this.

> transform.foreach is changed to take multiple handles by making target variadic and iterates over all handles at once by "zipping" the lists associated with the handles. The enclosed block would take as many arguments as there are targets.

Nit: make sure to define behavior when lengths don't match, e.g. `zip_shortest` vs. `zip` (produces failure on mismatch). Making it visible in the syntax is also nice.

> @ftynse, you mentioned an approach whereby the transform dialect would itself be subject to rewrites. Do you have a reference for us to look at?

I don't think something was committed upstream because there was no pressing need for that. However, transform dialect is a just a dialect. One should be able to write a pass based that applies patterns rewriting transform dialect operations quite easily. One of these patterns may expand a hypothetical high-level op into the actual implementation as proposed above.

muneebkhan85 commented:
@ftynse Thanks for the prompt feedback on this. So, @rolfmorel and I will work on these three patches separately and submit in separate PRs

muneebkhan85 commented:
@ftynse Thanks for the prompt feedback on this. 

> SGTM in principle. I'm not quite sure how this tiling would work for multiple dimensions though.

It should work the same way as it does for multitile_sizes where each dimension is tiled separately.
The above example from @rolfmorel only shows applying continuous tiles for dimension 0. The same
could be repeated for dimension 1 after utilizing replicate and split in the same way as is the case for
multi size tiling.

@rolfmorel and I will work on the proposed changes for `transform.structured.split` and
`transform.foreach` in separate PRs (that will refer to the discussion here) while repurposing this PR
to introduce `transform.structured.continuous_tile_sizes` op.

rolfmorel commented:
Here's an example of how, with the proposed interface, continuous tiling can be applied to two dimensions. This has the exponentially decreasing pattern of tile sizes for the inner dimension occurring within each tiling loop of the outer dimension (where these outer loops themselves have exponentially decreasing tile sizes):

```mlir
%linalg = transform.structured.match ops{["linalg.generic"]} in %payload // assume only one linalg matched
%tile_sizes_dim0, %split_points_dim0 = transform.structured.continuous_tile_sizes %linalg { dimension = 0, target_size = 9} : (!transform.any_op) -> (!transform.param<i64>, !transform.param<i64>)
%tile_sizes_dim1, %split_points_dim1 = transform.structured.continuous_tile_sizes %linalg { dimension = 1, target_size = 7} : (!transform.any_op) -> (!transform.param<i64>, !transform.param<i64>)
%linalg_dim0_splits = transform.structured.split %linalg after %split_points_dim0 { dimension  = 0, multiway }
transform.foreach %linalg_dim0_splits, %tile_sizes_dim0 {
  ^bb1(%linalg_dim0_split: !transform.any_op, %tile_size_dim0: !transform.any_param):
  %linalg_dim0_split_tiled, %tile_loop = transform.structured.tile_using_for %linalg_dim0_split [%tile_size_dim0]
  %linalg_dim1_splits = transform.structured.split %linalg_dim0_split_tiled after %split_points_dim1 { dimension = 1, multiway }
  transform.foreach %linalg_dim1_splits, %tile_sizes_dim1 {
    ^bb2(%linalg_dim1_split: !transform.any_op, %tile_size_dim1: !transform.any_param):
    transform.structured.tile_using_for %linalg_dim1_split [0, %tile_size_dim1]
    transform.yield
  }
  transform.yield
}
```

In the above, the `multiway` attribute tells `transform.structured.split` to interpret the vector of tile sizes associated with the `split_points` argument as a list of multiple points to split the target linalg along (rather than the handles for `target` and `split_points` being co-indexed with just one split point for each linalg).

This pattern naturally extends to higher dimensions.

ftynse commented:
Ok, this design is fine with me.

MaheshRavishankar commented:
> > High level comment, why cant this just be an after tile and fuse peeling transformation. What is happening here seems to be peeling applied repeatedly.
> 
> An issue here is that the static tile size is created using AffineMin at the time of creating the loop nests. This makes the issue more complicated than just having to create new tail-loops and clone regions from the original loop into it.

Sorry for the delay. I am still not convinced this is the easiest way to doing this. Looking at the description, you are trying to split full tiles and partial tiles (and then applying that recursively). As easier approach to me is to do something like this

```
%result = <linalg_op>(%operand1, %operand2, ...)
```

you can first split into two
```
%operand0_slice = tensor.extract_slice %operand0
%operand1_slice = tensor.extract_slice %operand1
....
%result_slice = <linalg_op>(%operand0_slice, %operand1_slice,...)
...
%operand0_remaining = tensor.extract_slice %operand0
%operand1_remaining = tensor.extract_slice %operand1
....
%result_remaining = <linalg_op>(%operand0_remaining, %operand1_remaining)
....
%result = tensor.insert_slice %result_slice into ....
%result1 = tensor.insert_slice %result_remaining  into ...
```

Now you can tile the individual ops. This does not need any changes to the  Tiling implementation itself, and is also modular. You can take the second linalg op and apply the same procedure repeatedly to get all the different loops. That would be upto the caller to do (or wrapped in a helper function), but the core logic is just this one step.

muneebkhan85 commented:
> > > High level comment, why cant this just be an after tile and fuse peeling transformation. What is happening here seems to be peeling applied repeatedly.
> > 
> > 
> > An issue here is that the static tile size is created using AffineMin at the time of creating the loop nests. This makes the issue more complicated than just having to create new tail-loops and clone regions from the original loop into it.
> 
> Sorry for the delay. I am still not convinced this is the easiest way to doing this. Looking at the description, you are trying to split full tiles and partial tiles (and then applying that recursively). As easier approach to me is to do something like this
> 
> ```
> %result = <linalg_op>(%operand1, %operand2, ...)
> ```
> 
> you can first split into two
> 
> ```
> %operand0_slice = tensor.extract_slice %operand0
> %operand1_slice = tensor.extract_slice %operand1
> ....
> %result_slice = <linalg_op>(%operand0_slice, %operand1_slice,...)
> ...
> %operand0_remaining = tensor.extract_slice %operand0
> %operand1_remaining = tensor.extract_slice %operand1
> ....
> %result_remaining = <linalg_op>(%operand0_remaining, %operand1_remaining)
> ....
> %result = tensor.insert_slice %result_slice into ....
> %result1 = tensor.insert_slice %result_remaining  into ...
> ```
> 
> Now you can tile the individual ops. This does not need any changes to the Tiling implementation itself, and is also modular. You can take the second linalg op and apply the same procedure repeatedly to get all the different loops. That would be upto the caller to do (or wrapped in a helper function), but the core logic is just this one step.

@MaheshRavishankar Thank you for the comments. After some deliberation we have decided to take another route for this without requiring any changes to the tiling logic. Please see the discussions here

https://github.com/llvm/llvm-project/pull/82792#issuecomment-2011851728

together with the example posted here

https://github.com/llvm/llvm-project/pull/82792#issuecomment-2015240405

The idea is to create an op `continuous_tile_sizes` that -- when provided a linalg op, dimension to tile and tile size -- would return as result 1) a list of `tile sizes`, and 2) a list of `split points` in a pattern similar to `multitile_sizes` except that each of these lists would contain more than one element. Then, using a multi-way split, we could use the existing tiling op to achieve  continuous tiling.

muneebkhan85 committed:
[mlir][Transform] Add continuous tiling to Transform dialect

Add continuous tiling op `structured.continuous_tile_sizes`
to the transform dialect that returns as result (1) a list of
exponentially diminishing tile sizes, and (2) a list of chunk
sizes -- along the specified dimension of the target --
where the corresponding tile sizes from (1) can be applied.
The list of chunk sizes from (2) cover the entire iteration
space along the given dimension of the target.

muneebkhan85 committed:
[mlir][Transform] Add support for multiway split in SplitOp

Add functionality that enables SplitOp to do a multiway split of
a traget op along a given dimension. With multiway attribute,
SplitOp takes a list of chunk sizes and applies it to a single
target along the given dimension to generate multiple
structured ops extracted from the target.

muneebkhan85 committed:
[mlir][test] Test multiway SplitOp

Tests SplitOp for multiway splitting of a structured op using
the result of `continuous_tile_sizes` to specify the
chunk sizes that the target's specified dimesion should be
split into.

muneebkhan85 commented:
@ftynse @MaheshRavishankar I have re-purposed this PR in line with our earlier discussion here and my latest commits (1) introduce a continuous tiling operation (`structured.continuous_tile_sizes`) that computes the various tile sizes and split points along the given dimension of a traget linalg op, and (2) enabling multiway splitting in SplitOp `transform.structured.split` that allows for the target op to be split at multiple points along the given dimension. (3) I have also inlcuded a test that also serves as an example on how to use results from `structured.continuous_tile_sizes` with multiway split op.

The third part of the proposal was to modify `transform.foreach`  as described in the comment above, also accessible here 

https://github.com/llvm/llvm-project/pull/82792#issuecomment-2011851728

@rolfmorel is currently working with that and will submit it as a separate PR.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+      llvm::to_vector(state.getPayloadOps(getTarget()));
+
+  if (!llvm::hasSingleElement(targetOps)) {
+    return emitDefiniteFailure() << "requires exactly one target (got "
```
Let's use silenceable failures unless the payload IR was modified.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+
+  auto target = dyn_cast<LinalgOp>(*targetOps.begin());
+
+  OpBuilder builder(target.getContext());
```
Use the provided `rewriter` instead.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+    auto tileSizeTripCountPairs =
+        llvm::zip_equal(spec->tileSizes, spec->tripCounts);
+
+    for (auto [idx, pair] : llvm::enumerate(tileSizeTripCountPairs))
```
`for (auto &&[idx, tileSize, tripCount] : llvm::enumerate(spec->tileSizes, spec->tripCounts)`

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+      splitPoints.push_back(std::get<0>(pair) * std::get<1>(pair));
+
+    auto makeI64AttrsFromI64 = [&](ArrayRef<int64_t> values) {
+      return llvm::to_vector(
+          llvm::map_range(values, [&](int64_t value) -> Attribute {
```
`llvm::map_to_vector`, we might also have a helper for this somewhere already, maybe @MaheshRavishankar knows.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+  SmallVector<Value> splitPoints;
+  Value splitPoint;
+  for (auto [idx, pair] : llvm::enumerate(tileSizeTripCountPairs)) {
+    splitPoint = apply(s0 * s1, {std::get<0>(pair), std::get<1>(pair)});
```
Why go through affine here? It's not good at simplifying multiplication.

muneebkhan85 replied:
`Affine Apply` is needed to generate the IR to compute the chunk sizes dynamically. This is need to support the case when bounds are not known. I'm not sure if there's another way to support that.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+    splitPoints.push_back(splitPoint);
+  }
+
+  auto makeOpFromValue = [&](ArrayRef<Value> values) {
```
```suggestion
  auto getDefiningOps = [&](ArrayRef<Value> values) {
```

ftynse replied:
It doesn't actually _make_ the op.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+  }
+
+  auto makeOpFromValue = [&](ArrayRef<Value> values) {
+    return llvm::to_vector(
+        llvm::map_range(values, [&](Value value) -> Operation * {
```
map_to_vector

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2581,6 +2643,157 @@ DiagnosedSilenceableFailure transform::TileReductionUsingForallOp::applyToOne(
+        }));
+  };
+
+  transformResults.set(cast<OpResult>(getTileSizes()),
+                       makeOpFromValue(spec->tileSizes));
+  transformResults.set(cast<OpResult>(getSplitPoints()),
+                       makeOpFromValue(splitPoints));
```
FYI, we have value handles. They are rather underused and I won't push for them here to maintain consistency, but it would make this cleaner.

muneebkhan85 replied:
This is a good suggestion, but skipping for consistency.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -107,6 +107,139 @@ static void emitIsPositiveIndexAssertion(ImplicitLocOpBuilder &b,
+  Value targetSizeValue =
+      getValueOrCreateConstantIndexOp(builder, loc, targetSize);
+
+  // Find the trip count of the iteration space dimension for which the tile
+  // sizes are computed.
+  SmallVector<OpFoldResult> allShapes =
+      op.createFlatListOfOperandDims(b, b.getLoc());
+  AffineMap shapesToLoops = op.getShapesToLoopsMap();
+  SmallVector<OpFoldResult> loopRanges =
+      makeComposedFoldedMultiResultAffineApply(b, op.getLoc(), shapesToLoops,
+                                               allShapes);
+
+  Value loopRange =
+      getValueOrCreateConstantIndexOp(b, op.getLoc(), loopRanges[dimension]);
```
Can we use/adapt `TilingInterface` for this?

muneebkhan85 replied:
I have adapted the dynamic case `computeContinuousTileSizes` to entirely use TilingInterface. For consistency, I have modified the static case `computeStaticContinuousTileSizes` to take a TilingInterface op, but it still relies on a casted LinalgOp for use internally.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
Note to self (@ftynse) : I haven't double checked the math here.

ftynse commented:
Thanks, this looks much cleaner. I have a couple of small changes and a question about using an interface.

muneebkhan85 commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -1409,16 +1409,27 @@ def SplitOp : Op<Transform_Dialect, "structured.split",
+    In this mode a single `target` op is split into multiple parts covering
+    the iteration space of the specified dimension. `static_split_point` and
+    `dynamic_split_point` in this case is a list of chunk sizes that the given
+    dimension should be split into. With `multiway` it produces two handles;
```
Changed to `chunk sizes` now with the description saying how it works when given just one value. The description also documents what happens when the chunk sizes do not cover the entire iteration space.

muneebkhan85 commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -1819,6 +1830,52 @@ def TileReductionUsingForallOp :
+        DeclareOpInterfaceMethods<TransformOpInterface>,
+        ReportTrackingListenerFailuresOpTrait]> {
+  let description = [{
+    This transform takes a linalg as target and a dimension and target size
```
Fixed.

muneebkhan85 commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2269,8 +2269,20 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
   SmallVector<Operation *> payload =
       llvm::to_vector(state.getPayloadOps(getTarget()));
+
+  bool isMultiwaySplit = getMultiway() ? true : false;
```
Don't know what I was thinking here. Fixed.

muneebkhan85 commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2269,8 +2269,20 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+  bool isMultiwaySplit = getMultiway() ? true : false;
+
+  if (isMultiwaySplit && !llvm::hasSingleElement(payload)) {
+    return emitDefiniteFailure() << "requires exactly one target when "
```
Fixed.

muneebkhan85 commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2305,57 +2319,105 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+    // Split a single target operation at multiple points.
+    SmallVector<Operation *> opList;
+    Operation *head, *tail;
+    for (const auto [idx, splitPoint] : llvm::enumerate(splitPoints)) {
```
Fixed.

muneebkhan85 commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2305,57 +2319,105 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+                                           << " does not exist in target op";
+        diag.attachNote(target->getLoc()) << "target op";
+        return diag;
+      }
```
I have refactored the code so that the common checks in the if/else are not duplicated and instead now rely on two lambda functions `checkStructuredOpAndDimensions` and `checkFailureInSplitting`. The code can not be entirely moved out of the loop as in the no-multiway case there can be several different payloads and each needs to be checked separately.

muneebkhan85 commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2305,57 +2319,105 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+          rewriter, cast<TilingInterface>(linalgOp.getOperation()),
+          getDimension(), splitPoint);
+
+      opList.push_back(head);
```
Fixed and Refactored. Uses the lambda function `checkFailureInSplitting` for doing the check.

muneebkhan85 commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -1409,16 +1409,27 @@ def SplitOp : Op<Transform_Dialect, "structured.split",
+    the first handle is a list of the multiple parts of the structured op
+    after splitting, where the target dimensions for each linalg op in the
+    list corresponds to the chunk sizes specfied in the input split list.
+    The second handle is empty.
```
I have tested this out. It can surely be a single handle and can be split into multiple handles using split_handle, though there's a caveat that split_handle needs to specify the expected number of outputs (as list of output types) depending on the number of payloads it receives in the handle.

In the multitile-size example `mlir/test/Dialect/Linalg/multisize-tiling-full.mlir` the 2nd split op generates two lists that are used by the two subsequent `tile_using_for`. Creating a single output for the split op would mean merging these two lists and `split_handle` would then see a handle of 4 payloads. This complicates things a bit in the sense that some reworking of the usage and examples are needed. Keeping it to produce two outputs for now. I think this change would be good for a separate PR.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -107,6 +107,138 @@ static void emitIsPositiveIndexAssertion(ImplicitLocOpBuilder &b,
+                                               unsigned dimension,
+                                               unsigned targetSize) {
+
+  LinalgOp linalgOp = dyn_cast<LinalgOp>(op.getOperation());
```
There's no check that `linalgOp` is non-null here. Given that the code below assumes `op` is a Linalg op, I'd just take a `LinalgOp op` in this function and let the caller do the casting.

muneebkhan85 replied:
Fixed. Reverted changes here to use LinalgOp instead of TilingInterface.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -107,6 +107,138 @@ static void emitIsPositiveIndexAssertion(ImplicitLocOpBuilder &b,
+
+  // Find the trip count of the iteration space dimension for which the tile
+  // sizes are computed.
+  SmallVector<Range> loopRanges = op.getIterationDomain(builder);
```
There's already a call to `op.getIteartionDomain` above. Given that it may emit IR, let's not duplicate it and rather preserve the value from above.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -107,6 +107,138 @@ static void emitIsPositiveIndexAssertion(ImplicitLocOpBuilder &b,
+  // Find the trip count of the iteration space dimension for which the tile
+  // sizes are computed.
+  SmallVector<Range> loopRanges = op.getIterationDomain(builder);
+  Value loopRange = getValueOrCreateConstantIndexOp(b, op->getLoc(),
```
```suggestion
  Value loopRange = getValueOrCreateConstantIndexOp(b, loc,
```

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -107,6 +107,138 @@ static void emitIsPositiveIndexAssertion(ImplicitLocOpBuilder &b,
+  OpFoldResult tripCountSize = affine::makeComposedFoldedAffineApply(
+      b, b.getLoc(), s0.floorDiv(s1), {loopRange, targetSizeValue});
+
+  uint64_t tileSizeInt = *getConstantIntValue(targetSizeValue);
```
Is it guaranteed or at least documented somewhere that `targetSizeValue`, in the case when it's forwarded from another Value inside OpFoldResult given to this function, is indeed a constant integer?

muneebkhan85 replied:
`emitAssertions` check (enabled at function call) asserts that `targetSize` is a positive integer. I have now added a comment about it here too.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2286,103 +2299,168 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+    if (diag.isSilenceableFailure())
+      return diag;
+
+    for (const auto &&[idx, chunkSize] : llvm::enumerate(chunkSizes)) {
```
Nit: we don't do `const` on IR objects.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2286,103 +2299,168 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+    Operation *target = payload.front();
+
+    auto linalgOp = dyn_cast<LinalgOp>(target);
+    auto diag = checkStructuredOpAndDimensions(linalgOp, target->getLoc());
```
Nit: please expand `auto` unless the type is obvious from context or impossible to spell.

muneebkhan85 replied:
Fixed.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2286,103 +2299,168 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+    Operation *head, *tail;
+    Operation *target = payload.front();
+
+    auto linalgOp = dyn_cast<LinalgOp>(target);
```
The result of `dyn_cast` must be systematically checked for not being null. If the operation is expected to always be of a certain class, use `cast`.

muneebkhan85 replied:
`checkStructuredOpAndDimensions` is called right after `dyn_cast` and checks if linalgOp is null and emits an error.

ftynse replied:
Please document that. Or, better, have `checkStructuredOpAndDimensions` take `Operation *`.

muneebkhan85 replied:
Included a comment now.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2286,103 +2299,168 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+      if (!target)
+        break;
+
+      linalgOp = dyn_cast<LinalgOp>(target);
```
Why do we need to do this on every iteration?

muneebkhan85 replied:
`linalg::splitOp` returns two `TilingInterface` values as result, mapped to `head` and `tail`. Even though the target has once been checked before the loop for being `LinalgOp`, I do this repeated casting in every iteration just as a formality since what we get originally from `linalg::splitOp` in `tail` is a `TilingInterface`. I have now also changed the types of head and tail to `TilingInterface` to make it clear in the code as to what type of results we get from `linalg::splitOp`.

MaheshRavishankar reviewed (commented):
Just a fly-by comment. I cant say for sure, but please refrain from using any of the `linalg::tile*` methods and use the `scf::tileUsingSCG*` methods. The `linalg::` are being removed.

rolfmorel commented:
Here's a suggestion for a test case. It requires the pending change to `transform.foreach` to land first though (#93705).

```mlir
// RUN: mlir-opt --transform-interpreter --split-input-file %s | FileCheck %s

PAYLOAD_CONTAINING_ANY_LINALG_HERE

module attributes {transform.with_named_sequence} {
  transform.named_sequence @tile_dim0_continously(%linalg: !transform.any_op {transform.readonly}, %target_size: !transform.param<i64>) -> !transform.any_op, !transform.any_op {
    %tile_sizes, %split_points = transform.structured.continuous_tile_sizes %linalg descending_from %target_size { dimension = 0 } : (!transform.any_op, !transform.param<i64>) -> (!transform.param<i64>, !transform.param<i64>)
    %linalg_splits, %empty = transform.structured.split %linalg after %split_points { dimension  = 0, multiway } : ...
    %tiled_splits, %loops = transform.foreach %linalg_splits, %tile_sizes {
      ^bb1(%linalg_split: !transform.any_op, %tile_size: !transform.param<i64>):
      %tiled_linalg_split, %dim0_loop = transform.structured.tile_using_for %linalg_split [%tile_size] : ...
      transform.yield %tiled_linalg_split, %dim0_loop : ...
    }
    transform.yield %tiled_splits, %loops : ...
  }
  transform.named_sequence @__transform_main(%payload: !transform.any_op) {
    %linalg = transform.structured.match ops{["linalg.generic"]} in %payload 
    %target_size = transform.param.constant 9 : !transform.param<index>
    // NB: we would need a foreach here/in the named sequence if %linalg would associate multiple linalg.generics
    %4 = transform.include @tile_dim0_continously failures(propagate) (%linalg) : (!transform.any_op) -> !transform.any_op
    transform.yield 
  }
}
```

Note the above assumes that `transform.structured.continuous_tile_sizes` is able to take `target_size` as a param instead of passing it as an attribute. (Note as well that the above is not quite syntactically correct: e.g., some types are left as homework.)

## Demonstrating abstraction

The main point of interest regarding the test case is that it demonstrates how to do abstraction with the Transform dialect. That is, it shows how a higher-level transform can be implemented with the Transform dialect while the details of this implementation can be hidden. In this way, the higher-level transform's functionality is exposed basically* as if it were available like any other Transform op.

(*: Note that currently a named sequence cannot be parametric w.r.t. attribute arguments that its contained ops take as attributes and not as params. So each op that we would like to use in named sequences - and whose attribute arguments we like expose as arguments of the sequence - needs to be modified to take that attribute as a `transform.param` argument as well. This is not ideal.)

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2286,103 +2299,171 @@ SplitOp::apply(transform::TransformRewriter &rewriter,
+      if (!target)
+        break;
+
+      linalgOp = cast<LinalgOp>(target);
```
Nit: we already casted it above.

ftynse commented:
> The main point of interest regarding the test case is that it demonstrates how to do abstraction with the Transform dialect. That is, it shows how a higher-level transform can be implemented with the Transform dialect while the details of this implementation can be hidden. In this way, the higher-level transform's functionality is exposed basically* as if it were available like any other Transform op.
>
> (*: Note that (currently) a named sequence cannot be parametric w.r.t. attribute arguments that its contained ops take as attributes and not as params. So each op that we would like to use in named sequences - and whose attribute arguments we'd like to expose as arguments of the sequence - needs to be modified to take the relevant attributes as transform.param arguments as well. Additionally these attribute arguments need to be packaged up as params before being handed to the named sequence - see transform.param.constant right before transform.include up above. This is not ideal.)

I'm open to suggestions here. Note that, despite people writing a lot of it directly, transform dialect is still an IR so some choices can be made for a compiler to reason more easily about the IR, such as param constants being operations similarly to all other constants in MLIR.

rolfmorel commented:
> I'm open to suggestions here. Note that, despite people writing a lot of it directly, transform dialect is still an IR so some choices can be made for a compiler to reason more easily about the IR, such as param constants being operations similarly to all other constants in MLIR.

Yes, valid point - having the param constants out in front of `transform.include`s is mostly a stylistic concern and could even be useful if we were to do things to Transform IR itself.

The more serious issue we are seeing is that the current MO is for all ops whose attribute arguments we would like to abstract over are to be modified - op by op, argument by argument - to take these attributes as params as well. While the current approach works, it feels there's potentially a better mechanism for addressing this need.

As this is becoming off-topic for this PR, we will post on Discourse to start a discussion regarding better ways of dealing with the above.

muneebkhan85 committed:
[mlir][test] Test full continuous tiling

Introduce a full test case demonstrating the continuous tiling
idea end-to-end.

muneebkhan85 commented:
@ftynse I have introduced new test cases under `mlir/test/Dialect/Linalg/continuous-tiling-full.mlir` to demonstrate the full idea of how continuous tiling is applied. 

While the examples here test for static dimension sizes, we have encountered an issue for the dynamic case where dimension sizes are not known, i.e, `?x?`. It turns out that in this case, multiway splitOp generates an "additional" trailing linalg op whose dimensions would evaluate to zero at runtime (in case of generating `chunk_sizes` using `continuous_tile_sizes`). This can be attributed to `linalg::splitOp` creating a valid head and tail on the last `chunkSize`. This causes the foreach  #93705 to fail in this case as `linalg_splits` and `tile_sizes` are not the same length. We could have a fully working dynamic example, if the last element in `linalg_splits` could be "dropped" in some way. One way to do this is to add `zip_shortest` to `foreach` when it takes more than one handles. If you have another suggestion or preference, please let us know. 

Here's an example for the dynamic dimensions case. `linalg_splits` and `tile_sizes` do not have the same payload size (6 vs 5), which makes the `foreach` fail.

```
module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
    %0 = transform.structured.match ops{["linalg.matmul"]} in %arg1 : (!transform.any_op) -> !transform.any_op
    %tile_sizes, %chunk_sizes = transform.structured.continuous_tile_sizes %0 { dimension = 0, target_size = 9 } : (!transform.any_op) -> !transform.any_op
    %linalg_splits, %empty = transform.structured.split %0 after %chunk_sizes { dimension = 0, multiway } : !transform.any_op, !transform.any_op
    transform.foreach %linalg_splits, %tile_sizes : !transform.any_op, !transform.any_op {
    ^bb1(%linalg_split: !transform.any_op, %tile_size: !transform.any_op):
      %tiled_linalg_split, %dim0_loop = transform.structured.tile_using_for %linalg_split tile_sizes [%tile_size] : (!transform.any_op, !transform.any_op) -> (!transform.any_op, !transform.any_op)
      transform.yield
    }
    transform.yield
  }
}

func.func @continuous_tile_linalg_matmul(
  %arg0: tensor<?x?xf32>, %arg1: tensor<?x?xf32>, %arg2: tensor<?x?xf32>)
    -> tensor<?x?xf32> {
  %0 = linalg.matmul  ins(%arg0, %arg1: tensor<?x?xf32>, tensor<?x?xf32>)
                     outs(%arg2: tensor<?x?xf32>)
    -> tensor<?x?xf32>

  return %0 : tensor<?x?xf32>
}
```

muneebkhan85 committed:
[mlir][Transform] Add `zip_shortest` to foreach

Adds `zip_shortest` functionality to `foreach` so that when it takes
multiple handles of varying lengths - instead of failing - it shrinks
the size of all payloads to that of the shortest payload.

muneebkhan85 commented:
@ftynse I have included a commit that solves the aforementioned issue with dynamic bounds, by introducing `zip_shortest` attribute and functionality to the `foreach` op. I have also included a test case for dynamic bounds in  `mlir/test/Dialect/Linalg/continuous-tiling-full.mlir`.

If this looks like a viable solution to you, please help with merging this PR. Otherwise, please let me know if you have other suggestions.

ftynse commented on mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:
```diff
@@ -645,11 +648,12 @@ def ForeachOp : TransformDialectOp<"foreach",
   let regions = (region SizedRegion<1>:$body);
   let assemblyFormat =
-    "$targets `:` type($targets) (`->` type($results)^)? $body attr-dict";
+    "$targets attr-dict `:` type($targets) (`->` type($results)^)? $body";
```
How about turning the attribute into a keyword rather than putting the entire dictionary here, which looks rather awkward?

ftynse reviewed (commented):
I have a syntactic comment, but I'll land this and let you consider that comment in a separate PR.
